lib/documents.js

```diff
@@ -2716,6 +2716,253 @@ function urisOutputTransform(headers, data) {
+            }
+          }
+        }
+        readBatchArray.push(record);
+      }
+
+  if (readBatchArray.length !== 0) {
+    const batchdef = {uris: readBatchArray};
+    if(jobState.jobOptions.transform){
+      batchdef.transform = jobState.jobOptions.transform;
+    }
+    readDocs(jobState, true, batchdef,readBatchArray, readerId);
+  }
+}
+
+function readDocs(jobState, val, batchdef, readBatchArray, readerId){
+  readDocumentsImpl.call(jobState.docInstance, val, [batchdef])
+      .result((output) => {
+        jobState.docsReadSuccessfully+= readBatchArray.length;
+        if (jobState.jobOptions.onBatchSuccess) {
+          const progressSoFar = {
+            docsReadSuccessfully: jobState.docsReadSuccessfully,
+            docsFailedToBeRead: jobState.docsFailedToBeRead,
+            timeElapsed: (Date.now()-jobState.startTime)
+          };
+          const documents = readBatchArray;
+          jobState.jobOptions.onBatchSuccess(progressSoFar, documents);
+        }
+        if(jobState.lastBatch){
+          finishReader(jobState);
+        }
+        else {
+          onReadAllDocs(jobState, readerId);
+        }
+      })
+      .catch((err) => {
+        if(jobState.error){
+          finishReader(jobState);
+          return;
+        }
+        let errorDisposition = err;
+        if (jobState.jobOptions.onBatchError) {
+          const progressSoFar = {
+            docsReadSuccessfully: jobState.docsReadSuccessfully,
+            docsFailedToBeRead: readBatchArray.length,
+            timeElapsed: (Date.now()-jobState.startTime)
+          };
+          const documents = readBatchArray;
+          try {
+            errorDisposition = jobState.jobOptions.onBatchError(progressSoFar, documents, err);
+          } catch(err){
+            errorDisposition = err;
+          }
+        }
+
+        if(errorDisposition instanceof Error){
+          jobState.docsFailedToBeRead += readBatchArray.length;
+          jobState.error = errorDisposition;
+          jobState.stream.emit('error', errorDisposition);
+          finishReader(jobState);
+          return;
+        } else if(!Array.isArray(errorDisposition)) {
+          onReadAllDocs(jobState, readerId);
+          return;
+        }
+
+        switch(jobState.retryCount){
+          case 0: {
+            jobState.docsFailedToBeRead += readBatchArray.length;
+            jobState.error = new Error('RetryCount exceeded.');
+            jobState.stream.emit('error', jobState.error);
+            finishReader(jobState);
+            break;
+          }
+          default: {
+            jobState.retryCount--;
+            batchdef.documents = errorDisposition;
+            readDocs(jobState, val, batchdef,batchdef.documents, readerId);
+            break;
+          }
+        }
+      });
+}
+
+function finishReader(jobState) {
+  jobState.requesterCount--;
+  if(jobState.requesterCount === 0) {
+    console.log('inside finishReader '+jobState.docsReadSuccessfully)
```

Comment:
The console.log() should be removed after development of the API.

I would expect that the finalization code should end the output stream.